File: structmem.cpp
#include <initializer_list>
#include <iostream>
#include <memory>
#include <vector>

struct Seq {
    int id;
    std::shared_ptr<std::vector<int>> vals;
};

__attribute__((noinline))
auto upped_id(Seq seq) -> Seq {
    seq.id += 1;
    return seq;
}

__attribute__((noinline))
auto report(const Seq& seq) -> void {
    std::cout << seq.id << " has " << seq.vals->size() << "\n";
}

__attribute__((noinline))
auto report_changes(Seq a) -> void {
    auto b = upped_id(a);
    a.vals->push_back(0);
    report(a);
    report(b);
}

auto main() -> int {
    auto a = Seq {
        .id = 5,
        .vals = std::make_shared<std::vector<int>>(
            std::initializer_list { 6, 7 }
        ),
    };
    report_changes(a);
}

File: structmem.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Clone)]
struct Seq {
    id: i32,
    vals: Rc<RefCell<Vec<i32>>>,
}

#[no_mangle]
#[inline(never)]
fn upped_id(mut seq: Seq) -> Seq {
    seq.id += 1;
    seq
}

#[no_mangle]
#[inline(never)]
fn report(seq: &Seq) {
    println!("{} has {}", seq.id, seq.vals.borrow().len());
}

#[no_mangle]
#[inline(never)]
fn report_changes(a: Seq) {
    let b = upped_id(a.clone());
    a.vals.borrow_mut().push(0);
    report(&a);
    report(&b);
}

fn main() {
    let a = Seq {
        id: 5,
        vals: Rc::new(RefCell::new(vec![6, 7])),
    };
    report_changes(a.clone());
}
